### ISSUE TITLE
Detect non terminated fluent API call

### ISSUE THREAD

**Issue by ceki** [KengoTODA/errorprone-slf4j#143]
Hello, 

In relation with https://jira.qos.ch/browse/SLF4J-526 , I was wondering if it was possible to detect when a user forgets to call the log() method with SLF4J 2.0 fluent API?

Currently, the API works like this:
```
logger.atDebug().addArgument(12").log("Current temperature {}");
```
The last call contains the log and the message. Currently, the API is designed so that the log call always contains a message. This is intended to prevent the user from forgetting to call `log()`.

However, if there was a way to detect this error, we could change the API to something more pleasant like:
```
logger.atDebug().message("Current temperature {}").addArgument(12").log();
```
For sake of completeness, here is a link to [LoggingEventBuilder](https://www.slf4j.org/apidocs/org/slf4j/spi/LoggingEventBuilder.html)

Do you think errorprone-slf4j could be modified to support this use case?

**Comment by Stephan202** [comment]
It sounds like this request can be fulfilled using a new [`AbstractReturnValueIgnored`](https://github.com/google/error-prone/blob/master/core/src/main/java/com/google/errorprone/bugpatterns/AbstractReturnValueIgnored.java) subclass: the result of a `LoggingEventBuilder`-returning expression should not be ignored. Basically a variation on something like [this](https://github.com/lhotari/reactor-error-prone/blob/master/reactor-error-prone/src/main/java/com/github/lhotari/reactor/errorprone/ReactiveStreamsPublisherReturnValueIgnored.java).

**Comment by ceki** [comment]
@Stephan202  I am not familiar with errorprone so I can't really tell. Is Is it applicable `AbstractReturnValueIgnored`  when `log()` returns `void` ? 

Perhaps there is a test for non-terminated collection stream API calls?

**Comment by Stephan202** [comment]
@ceki yep, `log()` returning `void` (or anything other than `AbstractReturnValueIgnored`) would make it work beautifully :)

> Perhaps there is a test for non-terminated collection stream API calls?

Yep, as part of the built-in `ReturnValueIgnored` check; see [here](https://github.com/google/error-prone/blob/dd0ee50fa7059197f3ea4562d50e6195eb8dd356/core/src/main/java/com/google/errorprone/bugpatterns/ReturnValueIgnored.java#L132-L137).

So what you're asking seems possible; let's see whether @KengoTODA is up for adding it to this library. But perhaps we can even just file a PR to upstream Error Prone. (I can help with that, though I'd rather not commit to a time line; very busy atm.)

**Comment by KengoTODA** [comment]
Hi @ceki :wave:, thanks for your hard work! It really makes the Java ecosystem more valuable and more reliable, I believe.

And thanks @Stephan202 for your help. I think that the built-in [ReturnValueIgnored rule](https://errorprone.info/bugpattern/CheckReturnValue) is enough for this case.

What we need to achieve goal is just annotating the fluent API with `@CheckReturnValue` (no matter JSR305 or errorprone annotations), then errorprone (and other code analysis tools) will warn SLF4J users when user made some mistakes.

This approach has a benefit which tailor made rules cannot provide: We can change the interface of fluent SLF4J API without worrying the implementation of rules. When we implement tailor made rules, we usually hold a database like [annotated JDK](https://github.com/typetools/jdk) on the side of rules, and it should be kept updated for future. It's possible but quite hard experience, so I want to avoid such implementation as much as possible.  
In other words, this `@CheckReturnValue`-based approach uses the code of SLF4J as a single-source-of-truth. Sounds attractive, or not? 😄 At least, until SLF4J v2 API becomes stable, it's better to use this approach. 

One known drawback is that, we need to introduce a dependency on the annotation library. It may bring some complexity and troublesome to the release workflow of SLF4J API.  
However, it's just annotations so we do not need them at runtime. This dependency will not make the API users' final package larger.


In short: I suggest annotating SLF4J API with the `@CheckReturnValue` annotation, instead of adding rules to errorprone itself nor this plugin.

**Comment by Stephan202** [comment]
Completely agree with using `@CheckReturnValue` :+1:. (And I wonder why this suggestion didn't occur to me yesterday; must not have been focussed :thinking:.)

For an example of another popular library using this approach, see Mockito's [`Stubber`](https://github.com/mockito/mockito/blob/f32973d4e709830a5a9fb19c54f32659023d2b03/src/main/java/org/mockito/stubbing/Stubber.java#L43), which uses [this annotation](https://github.com/mockito/mockito/blob/6f9108b833323345bd926f10914224d3e0a11b80/src/main/java/org/mockito/CheckReturnValue.java).

**Comment by KengoTODA** [comment]
> For an example of another popular library using this approach, see Mockito's Stubber, which uses this annotation.

Ah, this approach is awesome, it needs no dependency. :+1:

**Comment by KengoTODA** [comment]
Created a PR for this idea. Let's talk with code for more discussions. :rocket:

https://github.com/qos-ch/slf4j/pull/279

**Comment by KengoTODA** [comment]
Will be resolved by SLF4J v2.0.4.

https://jira.qos.ch/browse/SLF4J-567
